### ISSUE TITLE
INIT_NET issues

### ISSUE THREAD

**Issue by andressbarajas** [KallistiOS/KallistiOS#617]
I was playing around with @darcagn's [httpd-ack-next repo](https://github.com/darcagn/httpd-ack-next/tree/testing) and I noticed that when using the INIT_NET flag in KOS_INIT_FLAGS(INIT_DEFAULT | INIT_NET); it would show a black screen for 45-60 seconds before showing the main screen.  Looking at the [code](https://github.com/KallistiOS/KallistiOS/blob/f92103258b6d0943bc657de1199eb9a5d8b4087d/kernel/net/net_core.c#L185) I found that KOS was trying to establish a DHCP connection which Im not setup for so it takes 45-60 secs to fail before it goes to showing the main screen of the repo.  By updating this portion of net_init() from

```
if(net_default_dev) {
    /* Did we get a requested IP address? this normally happens over dcload-ip. */
    if(ip) {
        rv = net_dhcp_request(ip);
        if(rv < 0) {
            dbglog(DBG_DEBUG, "Failed to acquire the specified IP with DHCP\n");

            /* If that fails, set the address manually. Although gateway
               and dns etc. will also manually need setting */
            net_ipv4_parse_address(ip, net_default_dev->ip_addr);
            rv = 0;
        }
    }
    /* We didn't get a requested IP address, if we don't already have one
       set, then do so via DHCP. */
    else if(!net_default_dev->ip_addr[0])
        rv = net_dhcp_request(0);
}
```
to this:
```
if(net_default_dev) {
    /* Did we get a requested IP address? this normally happens over dcload-ip. */
    if(ip) {
            /* Set the address manually. Although gateway
               and dns etc. will also manually need setting */
            net_ipv4_parse_address(ip, net_default_dev->ip_addr);
            rv = 0;
        }
    }
    /* We didn't get a requested IP address, if we don't already have one
       set, then do so via DHCP. */
    else if(!net_default_dev->ip_addr[0])
        rv = net_dhcp_request(0);
}
``` the main screen shows up within 2-3 seconds of upload using dcload-serial.  I have a question (as a networking noob):

1. Why do we do a call to net_dhcp_request(ip); when given an IP when it can also be set manually? For people that are not setup for DHCP, it blocks the main thread and we have 45-60 seconds of black screen before it times out and sets it manually.

Tagging experts and people Ive talked with this about: @Kazade @darcagn @gyrovorbis

**Comment by andressbarajas** [comment]
Also according to the documentation for net_init(uint32 ip): "To auto-detect the IP address to assign to the default device (i.e, over DHCP or from the flashrom on the Dreamcast), pass 0 as the IP parameter".  So when the IP is provided we should be setting it manually like I do in my update above.

**Comment by Kazade** [comment]
Fwiw this was problematic when developing Driving Strikers - I have to initialise the networking much later to avoid hanging on boot. But I can't think of another option really unless we can somehow make the initialisation of networking asynchronous...

**Comment by QuzarDC** [comment]
There are two different types of dhcp requests happening here. If you already have your own IP and you want to use it on the network, the first request asks the dhcp server (router) if it can be used on the network. It might be that it's not in a valid range or already in use or similar. The second kind with just '0' is saying 'I don't have an IP that I want so anything'll do'.

So if you have an IP configured through whatever means, even if it's got a correct gateway and such, there's no guarantee that it'll be allowed by the dhcp server of the network you're connecting to. If you try to use it without confirming with the dhcp server you could conflict with another device on the network.

So we can't just skip that part. What could be done though is to make sure it's async as possible.

**Comment by QuzarDC** [comment]
I think the thing to do here is to take a dive in profiling where any bottlenecks are deeper in the process of making a dns request on the DC side. It's likely that it would help reveal some of the deeper issues in the network stack as in #116 and would be a good focused starting point for that.

I've marked this for v2.2.0, and will keep it in mind.

**Comment by gyrovorbis** [comment]
> Fwiw this was problematic when developing Driving Strikers - I have to initialise the networking much later to avoid hanging on boot. But I can't think of another option really unless we can somehow make the initialisation of networking asynchronous...

That's what we really need, is to make it asynchronous, now that we have nice shit in KOS like worker threads and deferring interrupts.... We could basically make whatever API function that you call that actually requires network connectivity block if it hasn't finished yet, or, if it has, just never even need to block... just like C++'s `std::async`...

...Only darc told me that the thing `INIT_NET` is blocking waiting to get populated, that he needs access to is *a freaking global variable*!!! WHYYYYYY!? The worst API decision ever... we have no control over that.... lol. so yeah, maybe a band-aid won't be so easy here.

**Comment by darcagn** [comment]
> I might be wrong, but I think if dcload passed an IP address, then [we make a dhcp request asking to reserve that IP](https://github.com/KallistiOS/KallistiOS/blob/395d99a9421aed367d7d277065080301c1a1d54d/kernel/net/net_core.c#L184) - and this will then update the gateway and DNS (I'm pretty sure that's how I coded it 😅 )

You are correct, but in the case of @andressbarajas, he has his Dreamcast connected to his computer through a separate subnet that doesn't have a DHCP server, and is statically defining his IP address on the dcload-ip disc. dcload-ip is passing that IP along and KOS attempts to update the gateway and DNS server through DHCP and gets blocked at that point in time. He doesn't have anything set in his flashrom, I'm not sure if KOS would bypass this if he had a static setting for gateway/DNS set in the flashrom, but at a glance it seems like it would still try to get it from DHCP when passed an IP from dcload-ip, I think.

**Comment by QuzarDC** [comment]
> You are correct, but in the case of @andressbarajas, he has his Dreamcast connected to his computer through a separate subnet that doesn't have a DHCP server, and is statically defining his IP address on the dcload-ip disc. dcload-ip is passing that IP along and KOS attempts to update the gateway and DNS server through DHCP and gets blocked at that point in time. He doesn't have anything set in his flashrom, I'm not sure if KOS would bypass this if he had a static setting for gateway/DNS set in the flashrom, but at a glance it seems like it would still try to get it from DHCP when passed an IP from dcload-ip, I think.

Even if all the settings were stored in the flashrom, the appropriate thing to do would still be to request from DHCP. The *only* time that wouldn't be the case would be in the dcload-ip scenarios where dcload-ip has already obtained a lease and KOS should be informed we have an existing lease to be used.

**Comment by darcagn** [comment]
> Even if all the settings were stored in the flashrom, the appropriate thing to do would still be to request from DHCP. The _only_ time that wouldn't be the case would be in the dcload-ip scenarios where dcload-ip has already obtained a lease and KOS should be informed we have an existing lease to be used.

Could you elaborate on why you believe that to be the case? As far as I'm aware, that is not how any other platform handles static IP settings, and I don't understand how the behaviors would work. Why should users who set static IP settings be forced to wait for a DHCP timeout when they explicitly chose not to use DHCP? And why? Is this supposed to give the DHCP server an opportunity to veto the user's explicit static IP setting?

**Comment by QuzarDC** [comment]
Oh! If stored settings are intended to override DHCP, then that's fine. I was in the context of thinking that the stored settings were just the last ones provided. If all of them presume that you're only entering after unselecting a 'via dhcp' option or similar ways where you're choosing the 'manual' way, then that makes perfect sense too.

**Comment by QuzarDC** [comment]
> > or the general case I'd say it should likely also block
> 
> I don't really agree with this. It really only makes sense to me for small utilities and examples. I think it should work like commercial games do -- the program starts up without networking, and the developer should call functions to initialize networking when the time is appropriate. Of course, we should provide a simple "just connect with defaults" simplified function.

Well, sure that's a fine approach to take too. I missed any hint in your previous comment that the suggestion was to simply not auto-init networking though, and my description only makes sense in the current state where we're doing that. *If* networking is being automatically initialized pre-main for the user, then there should be assurance that like every other automatically initialized system it's ready to use when called.

**Comment by darcagn** [comment]
> Would need to review the code some but that INIT_NET is only meant to auto-initialize networking. If not set, you should be able to net_init on your own when you want. Looking around that seems to still be right (but maybe it's disabling networking generally somewhere too?). 

Yeah this used to be the case (I used to do it without `INIT_NET` myself), and looking at the code it still _seems_ to be the case, but when I was trying the other night, and simply calling `net_init(0)` myself, I couldn't get a connection. There's [only one place](https://github.com/KallistiOS/KallistiOS/blob/1e23f16044012c948bf2c04a984fc19081238a6c/kernel/arch/dreamcast/fs/fs_dcload.c#L474) that assumes `INIT_NET` is set if networking is on, and that's for dcload-ip and it would be blocking anyway. So I'm not sure why I saw that behavior. I'll have to take a look again soon.

**Comment by DC-SWAT** [comment]
> Would need to review the code some but that `INIT_NET` is only meant to auto-initialize networking. If not set, you should be able to `net_init` on your own when you want. Looking around that seems to still be right (but maybe it's disabling networking generally somewhere too?). Pretty sure that it _is_ broken in the current state if using dcload-ip (perhaps this was also one of the conditions that fostered the notion that you couldn't do both net and dcload-ip) as I don't see a way for a user to swap the dbgio from dcload to dclsocket.

Calling `net_init()` isn't enough. If you don't use `INIT_NET` then you need to directly initialize all network devices first, otherwise they won't link to binary and initialize at all. And if you start to initialize the network manually, you will not have a single network device.
This is the modularity problem, which I have already spoken about somewhere.
